[flang] Add hlfir.char_trim operation

luporl opened the pull request:
Fortran character trim is currently lowered directly into a runtime
call, which makes it more complex to simplify expressions using it.

With this patch trim is first lowered into an hlfir.char_trim
operation, that is only later transformed into a runtime call.
This makes it easier to remove unnecessary calls to trim, as
proposed in #154593.

luporl committed:
[flang] Add hlfir.char_trim operation

Fortran character trim is currently lowered directly into a runtime
call, which makes it more complex to simplify expressions using it.

With this patch trim is first lowered into an hlfir.char_trim
operation, that is only later transformed into a runtime call.
This makes it easier to remove unnecessary calls to trim, as
proposed in #154593.

luporl commented:
Several tests from gfortran testsuite fail with this change.

vzakhari commented:
Thank you for the changes @luporl!  It looks good to me, and I am not sure why gfortran tests might be failing.

luporl committed:
Change result char type length to unknown

The runtime Trim function allocates the result, that may have a
different length than its input character string.

luporl commented:
Thanks for the review @vzakhari!

The issue was related to the return type of `hlfir.char_trim`.
As the runtime Trim function allocates the result and as it may have a different length than its input character string, it must always be a character with unknown length.

valerydmit reviewed (approved):
LGTM. Thank you!

tblah reviewed (approved):
LGTM, thanks for the fix

jeanPerier reviewed (approved):
LGTM, thanks!
